File: leetcode/325-maximum-size-subarray-sum-equals-k/main.go
package main

import "fmt"

/*
	Questions to ask:
	- what if the array is empty?
	- for [1,99,999], 999, return 1? becos there is one subarray, which has only one item 999, add up to 999?
*/

/*
	1st approach: prefix sum
	Time O(n^2)
	Space O(n)
	TLE
*/

/*
	2nd approach:
	- learned from others: https://leetcode.com/problems/maximum-size-subarray-sum-equals-k/discuss/77807/Clean-python-solution-one-pass
	- the basic idea is to store the previous sum in a hashtable
			e.g. key: previous sum, value: index of the previous sum
	- if currentSum - target in the hastable, the result is currentIndex - hastable[previous sum]
	Time O(n)
	Space O(n) hashtable
	24ms beats 100%
*/
func maxSubArrayLen(nums []int, k int) int {
	res := 0
	acc := 0
	ht := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		acc += nums[i]
		if acc == k {
			res = i + 1
		} else if _, x := ht[acc-k]; x {
			if i-ht[acc-k] > res {
				res = i - ht[acc-k]
			}
		}
		if _, x := ht[acc]; !x {
			ht[acc] = i
		}
	}
	return res
}

func main() {
	fmt.Println(maxSubArrayLen([]int{1, -1, 5, -2, 3}, 3))
	fmt.Println(maxSubArrayLen([]int{-2, -1, 2, 1}, 1))
	fmt.Println(maxSubArrayLen([]int{-2, -1, 2, 1, 100}, 100))
	fmt.Println(maxSubArrayLen([]int{-2, -1, 2, 100, 1}, 100))
	fmt.Println(maxSubArrayLen([]int{-2, -1, 2, 1000}, 99))
}

File: leetcode/325-maximum-size-subarray-sum-equals-k/main.py
"""
    1st approach: zero sum subarray
    - this question is fucking similar to leetcode523, 525, 560, 930, 1124, 1171
    - learned from others: https://leetcode.com/problems/maximum-size-subarray-sum-equals-k/discuss/77807/Clean-python-solution-one-pass
    - the basic idea is to store the previous sum in a hashtable
        e.g. key: previous sum, value: index of the previous sum
    - if currentSum - target in the hastable, the result is currentIndex - hastable[previous sum]

    ref:
    - https://www.youtube.com/watch?v=hLcYp67wCcM

    Time O(n)
    Space O(n) hashtable
    36ms beats 100%
"""


class Solution(object):
    def maxSubArrayLen(self, nums, k):
        res = 0
        pfs = 0
        ht = {}
        for i in range(len(nums)):
            pfs += nums[i]
            if pfs == k:
                res = i+1
            remain = pfs - k
            if remain in ht:
                res = max(res, i - ht[remain])
            if pfs not in ht:
                ht[pfs] = i
        return res


print(Solution().maxSubArrayLen([1, -1, 5, -2, 3], 3))
print(Solution().maxSubArrayLen([-2, -1, 2, 1], 1))
print(Solution().maxSubArrayLen([-2, -1, 2, 1, 100], 100))
print(Solution().maxSubArrayLen([-2, -1, 2, 100, 1], 100))
print(Solution().maxSubArrayLen([-2, -1, 2, 1000], 99))

print("-----")

"""
    2nd: same logic but just shorter
"""
class Solution:
    def maxSubArrayLen(self, nums: List[int], k: int) -> int:
        ht = {0:-1}
        pfs = 0
        res = 0
        for i in range(len(nums)):
            pfs += nums[i]
            if pfs-k in ht:
                j = ht[pfs-k]
                res = max(res, i - j)
            if pfs not in ht: 
                ht[pfs] = i
        return res